stetho/src/main/java/com/facebook/stetho/inspector/elements/android/ApplicationDescriptor.java

```diff
@@ -0,0 +1,198 @@
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+final class ApplicationDescriptor extends ChainedDescriptor<Application> {
+  private static final String TAG = "ApplicationDescriptor";
+
+  private final Map<Application, ElementContext> mElementToContextMap =
+      Collections.synchronizedMap(new HashMap<Application, ElementContext>());
+
+  private ElementContext getContext(Application element) {
+    return mElementToContextMap.get(element);
+  }
+
+  @Override
+  protected void onHook(Application element) {
+    ElementContext context = newElementContext();
+    context.hook(element);
+    mElementToContextMap.put(element, context);
+  }
+
+  @Override
+  protected void onUnhook(Application element) {
+    ElementContext context = mElementToContextMap.remove(element);
+    context.unhook();
+  }
+
+  @Override
+  protected NodeType onGetNodeType(Application element) {
+    return NodeType.DOCUMENT_NODE;
+  }
+
+  @Override
+  protected int onGetChildCount(Application element) {
+    ElementContext context = getContext(element);
+    return context.getActivitiesList().size();
+  }
+
+  @Override
+  protected Object onGetChildAt(Application element, int index) {
+    ElementContext context = getContext(element);
+    return context.getActivitiesList().get(index);
+  }
+
+  private static List<Activity> getAllActivitiesHackQuietly() {
+    try {
+      return getAllActivitiesHack();
+    } catch (Exception e) {
+      LogUtil.w(TAG, e, "Cannot retrieve list of Activity instances. UI inspection may not work!");
+      return Collections.emptyList();
+    }
+  }
+
+  // HACK: https://androidreclib.wordpress.com/2014/11/22/getting-the-current-activity/
+  // TODO: I'm unsure which version(s) of Android this works on
+  private static List<Activity> getAllActivitiesHack()
+      throws ClassNotFoundException, NoSuchMethodException, NoSuchFieldException, IllegalAccessException, InvocationTargetException {
+    Class activityThreadClass = Class.forName("android.app.ActivityThread");
+    Object activityThread = activityThreadClass.getMethod("currentActivityThread").invoke(null);
+    Field activitiesField = activityThreadClass.getDeclaredField("mActivities");
+    activitiesField.setAccessible(true);
+    Map activities = (Map)activitiesField.get(activityThread);
+    List<Activity> activitiesList = new ArrayList<Activity>(activities.size());
+
+    for (Object activityRecord : activities.values()) {
+      Class activityRecordClass = activityRecord.getClass();
+      Field activityField = activityRecordClass.getDeclaredField("activity");
+      activityField.setAccessible(true);
+      Activity activity = (Activity)activityField.get(activityRecord);
+      activitiesList.add(activity);
+    }
+
+    return activitiesList;
+  }
+
+  private ElementContext newElementContext() {
```

Comment:
Separate class please.  Try to be discrete with class responsibility, especially when hacks or backwards compatibility is involved.